### ISSUE TITLE
Store error flag in the Redis cache during inference

### ISSUE THREAD

**Issue by rsas** [google/souper#173]
When using ``cache_infer``, it would great to store a flag if an error has occurred during inference including the error message (e.g. ``Protocol error``). Similarly, ``cache_dump`` should extended to retrieve those erroneous optimizations and their error codes.

**Comment by jeroenk** [comment]
Could you explain why you want this? In my opinion such "optimisations" should not make it into the cache and souper should just abort when an error occurred.

**Comment by rsas** [comment]
My current workflow using souper is as follows: compile some code with ``-souper-no-infer`` to populate Redis with LHSs and then experiment with inference (e.g. synthesis) using the ``cache_infer`` tool only. Finally, I dump the successful optimizations using using the ``cache_dump`` tool and analyze them. For the unsuccessful cases I have no information why e.g. synthesis failed. It can be solver timeout, protocol error/segfault (bug in the code) or there is simply no optimization possible. Therefore, I think it would be helpful to store the error message in Redis for both retrieval and analysis of such cases. We are planning to start synthesis jobs with longer solver timeouts after the initial (fast) synthesis mode completed. Thus, we'll have to retrieve the "solver timeout" LHSs for this task too. Does it make sense?

**Comment by jeroenk** [comment]
The timeout and no optimization cases make sense; segfaults and protocol errors should be fixed on the spot, I think.

**Comment by rsas** [comment]
I agree, segfaults/protocol errors would be just a convenient thing to retrieve the failing LHS.
